### PR TITLE
Add IPv6 support for ipvs proxier

### PR DESCRIPTION
**Pull request by Mashimiao**
Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

**What this PR does / why we need it**:
Currently, ipvs just supports IPv4 address for endpoints and ingress, we should also support IPv6

**Release note**:
```release-note
None
```

### CODE REVIEW

**Comment by Mashimiao** [comment]
ping @smarterclayton  @vishh

**Comment by m1093782566** [comment]
Sorry, I just don't understand why you change:

utilproxy.RevertPorts(replacementPortsMap, proxier.portsMap)		

->

proxyutil.RevertPorts(replacementPortsMap, proxier.portsMap)

and move `/pkg/proxy/util/endpoints*.go` to `/pkg/proxy/util/utils*.go`?

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -47,7 +47,7 @@ import (
 	"k8s.io/kubernetes/pkg/features"
 	"k8s.io/kubernetes/pkg/proxy"
 	"k8s.io/kubernetes/pkg/proxy/healthcheck"
-	utilproxy "k8s.io/kubernetes/pkg/proxy/util"
+	proxyutil "k8s.io/kubernetes/pkg/proxy/util"
```
Comment (m1093782566) [inline pkg/proxy/iptables/proxier.go]: Why change this?

File: pkg/proxy/ipvs/proxier.go [hunk 2]
```diff
@@ -1000,7 +1002,7 @@ func (proxier *Proxier) syncProxyRules() {
 			"-A", string(kubeServicesChain),
 			"-m", "comment", "--comment", fmt.Sprintf(`"%s cluster IP"`, svcNameString),
 			"-m", protocol, "-p", protocol,
-			"-d", fmt.Sprintf("%s/32", svcInfo.clusterIP.String()),
+			"-d", proxyutil.ToCIDR(svcInfo.clusterIP),
```
Comment (m1093782566) [inline pkg/proxy/ipvs/proxier.go]: That's a nice finding.

File: pkg/proxy/iptables/proxier.go (hunk 1 above)
Comment (Mashimiao) [reply pkg/proxy/iptables/proxier.go]: Other imports like "k8s.io/apiserver/pkg/AAA/BBB", the short name always is AAABBB, so I think we may follow this order better.
So, according to them the short name of "k8s.io/apiserver/pkg/util/feature", should be `proxyutil`.

Comment (m1093782566) [reply pkg/proxy/iptables/proxier.go]: Hmmm, pkg/proxy/iptables/proxier.go has the same lines,

see: utilproxy "k8s.io/kubernetes/pkg/proxy/util"

File: pkg/proxy/util/endpoints.go [hunk 3]
```diff
@@ -1,51 +0,0 @@
-/*
```
Comment (m1093782566) [inline pkg/proxy/util/endpoints.go]: Does the movement make sense?

**Comment by Mashimiao** [comment]
On 10/13/2017 11:13 AM, DuJun wrote:
> and move |/pkg/proxy/util/endpoints*.go| to |/pkg/proxy/util/utils*.go|?

Because as function ToCIDR() does not only work for endpoints,
we can also use it for ingress.
So, I think just put them into endpoints.go is not suitable.
As utility functions, putting them into utils.go may be better.

**Comment by m1093782566** [comment]
> Because as function ToCIDR() does not only work for endpoints,
we can also use it for ingress.
So, I think just put them into endpoints.go is not suitable.
As utility functions, putting them into utils.go may be better.

I have tried to organize some groups of util functions which do the same thing into different files, such as conntrack.go, port.go... Possibly what you need is a better file name, e.g. ip.go...

File: pkg/proxy/ipvs/proxier.go [hunk 4]
```diff
@@ -1455,11 +1457,15 @@ func getLocalIPs(endpointsMap proxyEndpointsMap) map[types.NamespacedName]sets.S
 	for svcPort := range endpointsMap {
 		for _, ep := range endpointsMap[svcPort] {
 			if ep.isLocal {
-				nsn := svcPort.NamespacedName
-				if localIPs[nsn] == nil {
-					localIPs[nsn] = sets.NewString()
+				// If the endpoint has a bad format, proxyutil.IPPart() will log an
```
Comment (m1093782566) [inline pkg/proxy/ipvs/proxier.go]: That's is a good fix as well. Iptables and ipvs has huge code duplication, that's why I try to refactor them to consume more identical codes. xref: #52528

**Comment by Mashimiao** [comment]
On 10/13/2017 11:29 AM, DuJun wrote:
> I have tried to organize some groups of util functions which do the same thing into different files, such as conntrack.go, port.go... Possibly what you need is a better file name, e.g. ip.go...
If so, I can update the PR.
Does the new name ip.go works for you?

**Comment by m1093782566** [comment]
> Does the new name ip.go works for you?

Yes, it seems good.

**Comment by Mashimiao** [comment]
@m1093782566  updated, PTAL

**Comment by m1093782566** [comment]
/lgtm

BTW, please assign IPVS Proxier related PR to me for review in the future. I would love to review :)

Thanks!

**Comment by Mashimiao** [comment]
On 10/13/2017 02:11 PM, DuJun wrote:
> /lgtm
> 
> BTW, please assign IPVS Proxier related PR to me for review in the future. I would love to review :)
> 
> Thanks!
> 

OK, got it.

**Comment by Mashimiao** [comment]
ping @smarterclayton  @m1093782566  @vishh

**Comment by Mashimiao** [comment]
ping @smarterclayton  @m1093782566   @vishh

**Comment by rramkumar1** [comment]
@Mashimiao Are you still pursuing this PR? It probably needs a rebase.

**Comment by m1093782566** [comment]
/close

As changes are already there.

**Pull request closed without merging**